client/network/src/request_responses.rs

```diff
@@ -0,0 +1,866 @@
+								peer_id,
+								handler,
+								event: ((*protocol).to_string(), event),
+							})
+						}
+						NetworkBehaviourAction::ReportObservedAddr { address } => {
+							return Poll::Ready(NetworkBehaviourAction::ReportObservedAddr {
+								address,
+							})
+						}
+					};
+
+					match ev {
+						// Received a request from a remote.
+						RequestResponseEvent::Message {
+							peer,
+							message: RequestResponseMessage::Request { request, channel },
+						} => {
+							let (tx, rx) = oneshot::channel();
+
+							// Submit the request to the "response builder" passed by the user at
+							// initialization.
+							if let Some(resp_builder) = resp_builder {
+								// If the response builder is too busy, silently drop `tx`.
+								// This will be reported as a `Busy` error.
+								let _ = resp_builder.try_send(IncomingRequest {
+									origin: peer.clone(),
+									request_bytes: request,
+									answer: tx,
+								});
+							}
+
+							let protocol = protocol.clone();
+							self.pending_responses.push(Box::pin(async move {
+								// The `tx` created above can be dropped if we are not capable of
+								// processing this request, which is reflected as a "Busy" error.
+								if let Ok(response) = rx.await {
+									RequestProcessingOutcome::PendingResponse {
+										protocol, inner_channel: channel, response
+									}
+								} else {
+									RequestProcessingOutcome::Busy { peer, protocol }
+								}
+							}));
+
+							// This `continue` makes sure that `pending_responses` gets polled
+							// after we have added the new element.
+							continue;
+						}
+
+						// Received a response from a remote to one of our requests.
+						RequestResponseEvent::Message {
+							message:
+								RequestResponseMessage::Response {
+									request_id,
+									response,
+								},
+							..
+						} => {
+							let out = Event::OutboundFinished {
+								request_id,
+								outcome: Ok(response),
+							};
+							return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+						}
+
+						// One of our requests has failed.
+						RequestResponseEvent::OutboundFailure {
+							request_id,
+							error,
+							..
+						} => {
+							let out = Event::OutboundFinished {
+								request_id,
+								outcome: Err(error),
+							};
+							return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+						}
+
+						// Remote has tried to send a request but failed.
+						RequestResponseEvent::InboundFailure { peer, error } => {
+							let out = Event::InboundRequest {
+								peer,
+								protocol: protocol.clone(),
+								outcome: Err(InboundError::Network(error)),
+							};
+							return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
+						}
+					};
+				}
+			}
+
+			break Poll::Pending;
+		}
+	}
+}
+
+/// Error when registering a protocol.
+#[derive(Debug, derive_more::Display, derive_more::Error)]
```

Comment:
This crate uses `derive_more` *and* `thiserror`. Maybe we could eventually select one?